### ISSUE TITLE
Offix 2.0 Architecture

### ISSUE THREAD

**Issue by wtrocki** [aerogear/offix#532]
## Feature Request

The architecture of the Offix

<img width="615" alt="Screenshot 2020-06-24 at 12 26 40" src="https://user-images.githubusercontent.com/981838/85547784-24bb1e80-b616-11ea-8f3f-6c763748354a.png">

What we need to do is to apply approach without optimistic response but apply the current change instead to cache/store and then have way to revert it. 

CC @Eunovo

**Comment by Eunovo** [comment]
Isn't this hinging on work in #533?
Before I got cut off, @wtrocki was talking about the "mixed approach" to storing data to avoid the performance issues he explained. I'm sorry to ask for another explanation but I'm lost here.

**Comment by wtrocki** [comment]
I think let's implement non mixed approach. IN this what we propose is:

Have generic engine that works with scheduler and any type of client REST, OData, GraphQL that has offline queue, network and scheduler. Those concepts exist in any distributed system that involves version vectors for consistency. 

Then we can use this concepts inside datastore that will not have initially capabilities to work with generic use cases, but those will be enabled by the libraries that datastore will use under the hood.

So users will loose very little capabilities and datastore can be used both as:
- sidekick for some specific offline use cases with generic capabilities
This will mean that datastore API can be mixed with graphql queries 

- Main driver with very little generic capabilities - for this use case people will need to manually execute datastore updates after generic operation finishes

- Mixed approach will come next but it will require us having solid base for both Offix DataStore and Offix 2.0 - in both cases those are still in development so let's not bring that now

**Comment by wtrocki** [comment]
TL;DR - Rome wasn't built in one day. Let's build a foundation first!
